src/librustc_mir/shim.rs

```diff
@@ -613,7 +619,12 @@ impl<'a, 'tcx> CloneShimBuilder<'a, 'tcx> {
         self.block(vec![], TerminatorKind::Resume, true);
     }
 
-    fn tuple_shim(&mut self, tys: &ty::Slice<ty::Ty<'tcx>>) {
+    fn tuple_like_shim(&mut self, tys: &[ty::Ty<'tcx>], kind: AggregateKind<'tcx>) {
+        match kind {
```

Comment:
Is there a reason for this assertion? This method works just as well for structs and univariant enums.

Reply:
Indeed, but it won't work for arrays. I guess the `match` can be extended in the future if needed.